### DISCUSSION TITLE
[已解决] 如何替换定义在dimen.xml里面的浮点型数据

### DISCUSSION THREAD

**Discussion by PeanutMelonSeedBigAlmond** [LSPosed/LSPosed#1522] [Q&A] [answered]
运行环境

|Android版本|Android 11|
|:-:|:-:|
|LSPosed版本|1.6.5 (6272)|
|Magisk版本|23.0 (23000)|
|Riru版本|v26.1.4.r522.8b379cedb5|
---

我想替换掉系统资源中的定义（定义在`/system/framework/framework-res.apk`里面），这个数据在源代码里面是这样定义的

```xml
<item name="config_wallpaperMaxScale" format="float" type="dimen">1.10</item>
```

我自己构造了一个dimen定义在模块的dimen.xml里面

```xml
<item name="max_wallpaper_scale" type="dimen" format="float">1</item>
```
这是替换的逻辑

```kotlin
class WallpaperScaleHook:IXposedHookZygoteInit{
    override fun initZygote(startupParam: IXposedHookZygoteInit.StartupParam) {
        val res=XModuleResources.createInstance(startupParam.modulePath,null)
        kotlin.runCatching {
            XResources.setSystemWideReplacement(
                "android",
                "dimen",
                "config_wallpaperMaxScale",
                res.fwd(R.dimen.max_wallpaper_scale)
            )
        }.onSuccess {
            Utils.Log.i("替换系统资源成功：config_wallpaperMaxScale")
        }.onFailure {
            Utils.Log.e(it.toString())
        }
    }
}
```

模块作用域我已经勾选了`系统框架`，重启之后日志没有报错，也正常打出了

`[ 2022-01-09T23:33:02.475     1000:  1646:  1646 I/LSPosed-Bridge  ] [INFO]: 替换系统资源成功：config_wallpaperMaxScale`

但是我在我的另外一个测试app上使用

```kotlin
getContext().resources.getFloat(
    Resources.getSystem().getIdentifier("config_wallpaperMaxScale", "dimen", "android")
)
```

获取到的结果还是`1.1`

请问如何解决？我参考其他人的代码都是使用`XResources.DimensionReplacement()`作为第三个参数传进去，但是这种方法似乎只适用于替换尺寸

---

我尝试过

```kotlin
XResources.setSystemWideReplacement(
    "android",
    "dimen",
    "config_wallpaperMaxScale",
    1.0f
)
```

```kotlin
XResources.setSystemWideReplacement(
    "android",
    "dimen",
    "config_wallpaperMaxScale",
    XResources.DimensionReplacement(1f,TypedValue.TYPE_FLOAT)
)
```

```kotlin
val value=TypedValue()
Resources.getSystem().getValue(
    Resources.getSystem().getIdentifier("config_wallpaperMaxScale","dimen","android"),
    value,
    true
)
XResources.setSystemWideReplacement(
    "android",
    "dimen",
    "config_wallpaperMaxScale",
    XResources.DimensionReplacement(1f,TypedValue.TYPE_FLOAT)
)
```

但是结果都和上面一样

**Comment by yujincheng08** [comment]
系统资源不是所有 app 共享一个的，所以是你作用域没设置对。

**Reply by PeanutMelonSeedBigAlmond** [reply]
> 系统资源不是所有 app 共享一个的，所以是你作用域没设置对。

那意思就是说，要把引用到这个值的所有app全部添加到作用域里面？

**Reply by yujincheng08** [reply]
✔️

**Reply by PeanutMelonSeedBigAlmond** [reply]
> ✔️

我按照建议勾选了模块作用域为`系统框架`、`系统界面`、`My Application（测试应用）`，按照以下代码重新实现了逻辑

日志正常打印，但是依然没有作用

![image](https://user-images.githubusercontent.com/25266280/148694157-643e3462-2ebf-4483-8eb0-dbb55173cd85.png)

请问能给出修改建议吗，因为原来并没有做过hook系统资源的功能，所以并不知道怎么改

先谢谢了

**Reply by yujincheng08** [reply]
先试试在模块中替换完成之后马上获取资源看看有没有替换成功

**Reply by PeanutMelonSeedBigAlmond** [reply]
按照提示，我在日志输出**成功**之后马上调用

```kotlin
Utils.Log.i("替换系统资源成功：config_wallpaperMaxScale")
val res1=Resources.getSystem().getFloat(
    Resources.getSystem().getIdentifier(
        "config_wallpaperMaxScale",
        "dimen",
        "android"
    )
)
Utils.Log.i("Resources.getSystem().getFloat() => $res1")
```

没有替换成功，给出的结果为`res1=1.1`

---

Update: 使用Edxposed载入模块，在不设置模块作用域，对所有app都启用时，依然无法替换

**Reply by yujincheng08** [reply]
开个 issue。我修好了😊。

**Reply by yujincheng08** [reply]
试试 #1524

**Comment by PeanutMelonSeedBigAlmond** [comment, marked as answer]
#1524 没有问题




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年1月10日(星期一) 凌晨4:20
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [LSPosed/LSPosed] 如何替换定义在dimen.xml里面的浮点型数据 (Discussion #1522)





 
试试 #1524
 
&mdash;

**Reply by yujincheng08** [reply]
可以用了？可以的话开个新 issue。
